File: exercism/c/binary_search.c
#include "binary_search.h"

int *binary_search(int value, const int *arr, size_t length) {
  int a = 0;
  int b = length - 1;
  if (length == 0 || arr == NULL)
    return NULL;
  while (a <= b) {
    int c = a + (b - a) / 2;
    if (arr[c] == value)
      return (int *)&arr[c];
    if (arr[c] < value)
      a = c + 1;
    else
      b = c - 1;
  }
  return NULL;
}

File: exercism/haskell/BinarySearch.hs
module BinarySearch
  ( find
  ) where

import           Data.Array              hiding ( array )

find :: Ord a => Array Int a -> a -> Maybe Int
find array x =
  let (start, end) = bounds array
  in  if start > end then Nothing else go start end (mid start end)
 where
  mid :: Int -> Int -> Int
  mid a b = (a + b) `div` 2
  go :: Int -> Int -> Int -> Maybe Int
  go a b c | a >= b        = if array ! a == x then Just a else Nothing
           | array ! c < x = go (c + 1) b (mid c b)
           | array ! c > x = go a (c - 1) (mid c a)
           | otherwise     = Just c
